src/MPQEditor/MPQEditor.ts

```diff
@@ -0,0 +1,645 @@
+    let granularity = Array<string>(names.length);
+    granularity.fill(this._content[MPQData._defGranularityKey]);
+    this.updateLayers(names, quantization, granularity);
+  }
+
+  filterModelNodesBy(filter: string[]) {
+    this._modelNodes = this._allModelNodes?.filter(
+      (name) => filter.find((filterName) => name === filterName) === undefined
+    );
+  }
+
+  getAllModelNodes(): string[] | undefined {
+    return this._modelNodes;
+  }
+
+  updateLayers(names: string[], quantization: string[], granularity: string[]) {
+    if (!(MPQData._layersKey in this._content)) {
+      this._content[MPQData._layersKey] = [];
+    }
+    for (let i = 0; i < names.length; i++) {
+      let layer = {
+        name: names[i],
+        dtype: quantization[i],
+        granularity: granularity[i],
+      };
+      this._content[MPQData._layersKey].push(layer);
+    }
+    this.filterModelNodesbyContent();
+  }
+
+  updateSectionOfLayer(name: string, section: string, value: string) {
+    let layer = this._content[MPQData._layersKey].find(
+      (x: any) => x["name"] === name
+    );
+    if (layer) {
+      layer[section] = value;
+    }
+  }
+
+  getSelected(): Set<string> {
+    return this._selectedNodes;
+  }
+
+  clearSelection() {
+    this._selectedNodes.clear();
+  }
+
+  removeModelNodesFromLayers(names: any) {
+    names.forEach((name: any) => {
+      let foundIndex = this._content[MPQData._layersKey].findIndex(
+        (x: any) => x["name"] === name
+      );
+      if (foundIndex > -1) {
+        this._content[MPQData._layersKey].splice(foundIndex, 1);
+      }
+    });
+    this.filterModelNodesbyContent();
+  }
+
+  getVisqPath(): string {
+    return this._visqPath;
+  }
+
+  setVisqPath(path: string): void {
+    this._visqPath = path;
+  }
+}
+
+export class MPQEditorProvider implements vscode.CustomTextEditorProvider {
+  public static readonly viewType = "one.editor.mpq";
+  public static readonly fileExtension = ".mpq.json";
+
+  private _disposables: vscode.Disposable[] = [];
+  private _mpqDataMap: any = {};
+
+  public static register(context: vscode.ExtensionContext): void {
+    const provider = new MPQEditorProvider(context);
+    const registrations = [
+      vscode.window.registerCustomEditorProvider(
+        MPQEditorProvider.viewType,
+        provider,
+        {
+          webviewOptions: { retainContextWhenHidden: true },
+        }
+      ),
+      // Add command registration here
+      vscode.commands.registerCommand(
+        "one.editor.mpq.showFromDefaultExplorer",
+        (uri) => {
+          MPQEditorProvider.createMPQJson(uri);
+        }
+      ),
+      vscode.commands.registerCommand(
+        "one.editor.mpq.showFromOneExplorer",
+        (uri) => {
+          MPQEditorProvider.createMPQJson(uri);
+        }
+      ),
+    ];
+
+    registrations.forEach((disposable) =>
+      context.subscriptions.push(disposable)
+    );
+  }
+
+  public static createMPQJson(uri: vscode.Uri): void {
```

Comment:
I think this function can be ensured by unit test, in the same context as your conversation with @seanshpark

Reply:
ok

Reply:
@dayo09 
Please correct me if i'm wrong, but createMPQJson contains 
```
vscode.window
      .showInputBox
```
seems like it's not valid for unittests. Will it be ok to split it into testable parts?